File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        int[] arr = new int[26]; // store count of each 26 character
        int longest = 0; // will return longest length of substring with only 1 repeating character
        int max = 0; // Max count of character that appear the most so far
        int l = 0; // left pointer
        
        for (int r = 0; r < s.length(); r++) { // incrementing right pointer
            arr[s.charAt(r) - 'A']++; // counting everytime each character appear
            max = Math.max(max, arr[s.charAt(r) - 'A']); // updating the max count of character that appear the most
            
            if (r - l + 1 - max > k) {
                arr[s.charAt(l) - 'A']--;
                l++;
            }
            
            longest = Math.max(longest, r - l + 1);
        }
        
        return longest;
    }
}

File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        '''
        initialize hashmap
        l and r pointer 
        
        calculate the number of char need replace x
        if x <= k: we update the length of our max winleng, update the fre of the character at r pointer, and then update r pointer
        if x > k: shrink the length of the the windown length by shifting l pointer to the right
        
        return
        '''
        
        count = {}
        l = 0
        res = 0
        for r in range(len(s)):
            count[s[r]] = 1 + count.get(s[r], 0)
            while (r - l + 1) - max(count.values()) > k:
                
                count[s[l]] -= 1
                l += 1
            
            
            res = max(res, r - l + 1)
            
        
        return res